Kernel/Process.cpp

```diff
@@ -1540,17 +1540,117 @@ void Process::crash(int signal, u32 eip)
+    elf_file_header.e_ident[EI_PAD + 1] = 0;
+    elf_file_header.e_ident[EI_PAD + 2] = 0;
+    elf_file_header.e_ident[EI_PAD + 3] = 0;
+    elf_file_header.e_ident[EI_PAD + 4] = 0;
+    elf_file_header.e_ident[EI_PAD + 5] = 0;
+    elf_file_header.e_ident[EI_PAD + 6] = 0;
+    elf_file_header.e_ident[EI_NIDENT] = 16;
+    elf_file_header.e_type = ET_CORE;
+    elf_file_header.e_machine = EM_386;
+    elf_file_header.e_version = 1;
+    elf_file_header.e_entry = 0;
+    elf_file_header.e_phoff = sizeof(Elf32_Ehdr);
+    elf_file_header.e_shoff = 0;
+    elf_file_header.e_flags = 0;
+    elf_file_header.e_ehsize = sizeof(Elf32_Ehdr);
+    elf_file_header.e_shentsize = sizeof(Elf32_Shdr);
+    elf_file_header.e_phentsize = sizeof(Elf32_Phdr);
+    elf_file_header.e_phnum = m_regions.size() + 1; // +1 for NOTE section
+    elf_file_header.e_shnum = 0;
+    elf_file_header.e_shstrndx = SHN_UNDEF;
+
+    fd->write(reinterpret_cast<uint8_t*>(&elf_file_header), sizeof(Elf32_Ehdr));
+
+    // TODO: We need to generate appropriate PT_NOTE sections to store
+    // register information and process status stuff here. This is done
+    // by creating a PT_NOTE Phdr for every threa and filling
+    // in the appropriate fields.
+
+    // Dump all regions to file
+    for (auto& region : m_regions) {
+        Elf32_Phdr phdr;
+        auto& vmobj = region.vmobject();
+
+        phdr.p_type = PT_LOAD;
+        phdr.p_offset = offset;
+        phdr.p_vaddr = reinterpret_cast<uint32_t>(region.vaddr().as_ptr());
+        phdr.p_paddr = 0;
+
+        size_t mapped_page_count = 0;
+        for (size_t page_num = 0; page_num < region.page_count(); page_num++) {
+            PhysicalPage* page = vmobj.physical_pages()[page_num];
+            if (!page)
+                continue;
+
+            mapped_page_count++;
+        }
+
+        phdr.p_filesz = mapped_page_count * PAGE_SIZE;
+        phdr.p_memsz = region.page_count() * PAGE_SIZE;
+        phdr.p_align = 0;
+
+        phdr.p_flags = region.is_readable() ? PF_R : 0;
+        if (region.is_writable())
+            phdr.p_flags |= PF_W;
+        if (region.is_executable())
+            phdr.p_flags |= PF_X;
+
+        offset += phdr.p_filesz;
+
+        fd->write(reinterpret_cast<uint8_t*>(&phdr), sizeof(Elf32_Phdr));
+    }
+
+    // Let's actually dump all the memory into the core file
+    for (auto& region : m_regions) {
+        // Remap the region so that it's readable by us.
+        region.set_readable(true);
+        region.remap();
+
+        // Now let's walk the physical pages and see which ones are actually mapped.
+        auto& vmobj = region.vmobject();
+        for (size_t i = 0; i < region.page_count(); i++) {
+            PhysicalPage* page = vmobj.physical_pages()[region.first_page_index() + i];
+            if (!page)
+                continue;
+
+            fd->write(reinterpret_cast<uint8_t*>((region.vaddr().as_ptr() + (i * PAGE_SIZE))), PAGE_SIZE);
```

Comment:
It kind of looks like this is assuming that a region will always be mapped from it's first page to some page towards the middle? I'm not sure this design will reconstruct the process's memory properly if you have
```
[ mapped | not mapped | mapped ]
```
?
The mapped physical pages would be copied into the file adjacent to each other. Do you have to create additional PT_LOAD segments for regions like this? Maybe I'm overthinking it.

Reply:
Looking at a readelf -aW of [this core dump](https://pastebin.com/Xjn6YRgN), It kind of looks like some of the PT_LOAD segments overlap? Not sure what I'm looking at with the two R E load headers where the file size is much smaller than the memory size..

Reply:
> It kind of looks like this is assuming that a region will always be mapped from it's first page to some page towards the middle? I'm not sure this design will reconstruct the process's memory properly if you have
> 
> ```
> [ mapped | not mapped | mapped ]
> ```
> 
> ?
> The mapped physical pages would be copied into the file adjacent to each other. Do you have to create additional PT_LOAD segments for regions like this? Maybe I'm overthinking it.

I think this should be taken into account with regards to how memory is supposed to be copied, but I'll have to reread the spec. Actually, I'm not 100% sure how the core files are supposed to handle pages that are unmapped, I was under the impression whatever _was_ mapped was just dumped into a `PT_LOAD` section, skipping over anything that's not currently mapped. Any ideas?

Also sorry for the horribly late reply on this, uni really kicked my ass the past month.